### ISSUE TITLE
Support for whitespace-separated CSS classnames in classMap directive.

### ISSUE THREAD

**Issue by nicholas-wilcox** [lit/lit#3095]
### Which package is this a feature request for?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

Sometimes an element's `classList` needs to update reactively in such a way that multiple class names are added or removed based on the same boolean condition. This might apply to other CSS methodologies, but in my recent experience, this has come up when using atomic classes (from TailwindCSS).

In other words, when using the `classMap` directive, it would be desirable to be able to do the following:

```javascript
return html`
  <div
    class=${classMap({
      "foo bar": this.booleanProperty,
      "baz qux": !this.booleanProperty
    })}
  ></div>
`;
```

This currently produces an error, as `classMap` assumes that the keys in the `classInfo` argument to be valid DOM tokens, rather than potentially being space-separated strings. I have created a minimal example [on CodePen](https://codepen.io/nicholas-wilcox/pen/VwXZbpE) to demonstrate the use case.

**Note:** This pattern does not produce an error until classes are added/removed _after_ the first render. If this feature should not be added, then it might be best to explicitly discourage the pattern in the documentation, or in a console warning.

### Alternatives and Workarounds

Developers can expand the `classInfo` object so that every class name in is a separate object property, which would duplicate the code used for their shared boolean value.

Alternatively, there is the `classnames` package [on NPM](https://www.npmjs.com/package/classnames), which supports this pattern, while (to my knowledge) completely avoiding the `DOMTokenList` API.

**Comment by justinfagnani** [comment]
Interesting.

I think the devil is in the details here. What happens in cases of conflicts, like:

```ts
return html`
  <div
    class=${classMap({
      "foo bar": true,
      "bar baz": false,
    })}
  ></div>
`;
```

Last wins I presume?

**Comment by nicholas-wilcox** [comment]
I would agree with that. It is the same as the current behavior I've seen from `classMap`. I could add a test to check this in both the single-class and multi-class cases.

**Comment by timbomckay** [comment]
We just ran into this on our project. We use utility classes within our components. It'd be nice to not require repeating the conditional for every class that needs to be applied. I understand why the classMap was created this way, don't think you'd expect people to use utility classes or multiple classes from a conditional due to the encapsulation of the shadow-dom. It'd be nice for this flexibility to be added though.

Regarding the conflict case above I think that can go a multitude of ways and it's gonna be on the developer at the end of it. You can do last one wins or it could be that true/add wins. Either way would be on the developer to ensure it's behaving the way they expect, as long as it's documented as a warning.

Currently the way we're solving for this is like so:

```js
render() {
  const classes = ['foo bar', 'padded'];

  if (this.enabled) {
    classes.push('enabled active');
  }

  if (this.hidden) {
    classes.push('hidden', 'hide');
  }

  return html`
    <div class=${classes.join(' ')}>Classy text</div>
  `;
}
```

**Comment by justinfagnani** [comment]
I would be open to a PR for this. Last-in wins sounds fine, considering that should be the current behavior with collisions just from JS object semantics.

**Comment by nicholas-wilcox** [comment]
I did create a PR (#3096) way back when that I believe satisfies the criteria we've settled on. I prioritized making the code changes as light as possible, which means that it might not be as optimized as it can/should be. I also added a test that should probably be expanded and broken up.

**Comment by ChellappanRajan** [comment]
Any update on this issue?

**Comment by hanifmhd** [comment]
any update on this issue? classmap is not working at all if there is whitespace css classname

**Comment by louwers** [comment]
I tried to use this wrapper

```js
/**
 *
 * @param {Parameters<typeof classMap>[0]} classInfo
 */
function ci(classInfo) {
  /** @type {Record<string, string | boolean | number>} */
  const newClassInfo = {};
  for (const [classNames, value] of Object.entries(classInfo)) {
    for (const className of classNames.split(" ")) {
      newClassInfo[className] = value;
    }
  }
  return newClassInfo;
}
```

```html
          <button
            class="p-2 rounded-3 border
            ${classMap(
              ci({
                "bg-slate-50 hover:bg-slate-100": needsConfirmation.value,
                "bg-red": !needsConfirmation.value,
              })
            )}
```

But I am getting this error:

> Error: `classMap()` can only be used in the `class` attribute and must be the only part in the attribute.

I probably don't understand directives well enough. Any hints?

**Comment by louwers** [comment]
You can also write your own wrapper:

```js
/** @type {typeof classMap} */
function myClassMap(classInfo) {
  /** @type {Record<string, string | boolean | number>} */
  const newClassInfo = {};
  for (const [classNames, value] of Object.entries(classInfo)) {
    for (const className of classNames.split(" ")) {
      newClassInfo[className] = value;
    }
  }
  return classMap(newClassInfo);
}
```

**Comment by JorisAerts** [comment]
I created a fix for this, see: https://github.com/lit/lit/issues/4149#issuecomment-3460757097
